virtcontainers/kata_agent.go

```diff
@@ -1550,6 +1556,64 @@ func (k *kataAgent) handleEphemeralStorage(mounts []specs.Mount) []*grpc.Storage
 	return epheStorages
 }
 
+// handleHugePages handles hugepages storage by
+// creating a Storage from corresponding source of the mount point
+func (k *kataAgent) handleHugepages(mounts []specs.Mount, hugepageLimits []specs.LinuxHugepageLimit) []*grpc.Storage {
+	//Map to hold the total memory of each type of hugepages
+	optionsMap := make(map[string]string)
+	var HugePageSizeUnitList = []string{"B", "K", "M", "G", "T", "P"}
+
+	for _, hp := range hugepageLimits {
+		if hp.Limit != 0 {
+			k.Logger().WithFields(logrus.Fields{
+				"Pagesize": hp.Pagesize,
+				"Limit":    hp.Limit,
+			}).Info("hugepage request")
+			//example Pagesize 2MB, 1GB etc. The Limit are in Bytes
+			pageSize, _ := units.RAMInBytes(hp.Pagesize)
+			pageSizeStr := units.CustomSize("%g%s", float64(pageSize), 1024.0, HugePageSizeUnitList)
+			totalHpSize := fmt.Sprintf("%v", hp.Limit)
+			optionsMap[pageSizeStr] = totalHpSize
+		}
+	}
+
+	var hugepages []*grpc.Storage
+	for idx, mnt := range mounts {
+		if mnt.Type != KataLocalDevType {
+			continue
+		}
+		//HugePages mount Type is Local
+		if _, fsType, fsOptions, _ := utils.GetDevicePathAndFsTypeOptions(mnt.Source); fsType == "hugetlbfs" {
+			k.Logger().WithField("fsOptions", fsOptions).Debug("hugepage mount options")
+			//Find the pagesize from the mountpoint to use the right set of options
+			//pagesize is the last element in the mount option. Can it change ?
+			pagesizeOpt := fsOptions[len(fsOptions)-1]
```

Comment:
> this seems too magical. I'd rather assume it can change.

@egernst I can go ahead and change the logic here. However can you please take a look at the following comment once and share your thoughts - https://github.com/kata-containers/runtime/pull/3109#discussion_r553878031

Reply:
@egernst pushed a new iteration with logic for searching the option. ptal